HW2/P2/parallel_vector.pyx

```diff
@@ -70,23 +70,58 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+    cdef:
+        int idx, r
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+ 
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+    for r in range(repeat):
+      # IN HERE EVERYTHING IS PARALLELIZED AND WE MUST BE CAREFUL
+      # FOR DEADLOCKS AND DOUBLE LOCKS
+        for idx in prange(src.shape[0], nogil=True, num_threads=4):
+            if src[idx]<dest[idx]:
+                #THIS PART MAKES SURE THAT IT IS IN ORDER AND AVOIDS
```

Comment:
You could consolidate this large if/else block and avoid rewriting the same code by just using something along the lines of 'aquire(&(locks[min(src[idx],dest[idx])]))' etc. But other than that, this works great!